contracts/test/StakeTable.t.sol

```diff
@@ -530,4 +553,162 @@ contract StakeTable_Test is Test {
+    function testFuzz_SequencesOfEvents(
+        uint8[ARRAY_SIZE] memory events,
+        uint8[ARRAY_SIZE] memory rands
+    ) external {
+        BN254.G2Point[ARRAY_SIZE] memory registeredKeys;
+
+        // Tracks the indices corresponding to an active key
+        bool[ARRAY_SIZE] memory isKeyActive;
+
+        uint64 numRegistrations = 0;
+        uint64 numExits = 0;
+
+        for (uint256 i = 0; i < ARRAY_SIZE; i++) {
+            uint256 ev = bound(events[i], 0, 2);
+
+            bool exitRequestSuccessful = false;
+
+            if (ev == 0) {
+                // Registrations
+                bool res = handleRegistrations(i, rands, registeredKeys, isKeyActive);
+                if (res) {
+                    numRegistrations++;
+                }
+            } else if (ev == 1) {
+                // Exits
+                if (numRegistrations == 0) {
+                    break;
+                }
+                uint256 indexRegistration = bound(rands[i], 0, numRegistrations - 1);
+                bytes32 hashNode = stakeTable._hashBlsKey(registeredKeys[indexRegistration]);
+                (
+                    address sender,
+                    AbstractStakeTable.StakeType stakeType,
+                    uint64 balance,
+                    uint64 registerEpoch,
+                    uint64 exitEpoch,
+                ) = stakeTable.nodes(hashNode);
+
+                balance;
+                stakeType;
+
+                BN254.G2Point memory blsVK = registeredKeys[indexRegistration];
+
+                bool canExit = (stakeTable.currentEpoch() >= registerEpoch + 1) && (exitEpoch == 0);
+                if (canExit) {
+                    vm.prank(sender);
+                    bool res = stakeTable.requestExit(blsVK);
+                    assertTrue(res);
+                    numExits++;
+                    exitRequestSuccessful = true;
+
+                    // Invariants specific to a successful exit
+                    assertGe(stakeTable.firstAvailableExitEpoch(), stakeTable.currentEpoch() + 1);
+                    assertGe(stakeTable.numPendingExits(), 1);
+                } else {
+                    vm.prank(sender);
+                    vm.expectRevert();
+                    bool res = stakeTable.requestExit(blsVK);
+                    assertFalse(res);
+                }
+            } else {
+                // Advance epoch
+                // ev == 2
+                uint64 currentEpoch = lightClientContract.currentEpoch();
+                uint64 nextEpoch = currentEpoch + 1;
+                lightClientContract.setCurrentEpoch(nextEpoch);
+                assertEq(stakeTable.currentEpoch(), nextEpoch);
+            }
+
+            // Global invariants
+            assertLe(stakeTable.numPendingRegistrations(), stakeTable.maxChurnRate());
+            assertLe(stakeTable.numPendingExits(), stakeTable.maxChurnRate());
+            assertLe(numExits, numRegistrations);
```

Comment:
why is this an invariant?

Reply:
Because one needs to register before exiting and thus the number of exits is always lower or equal to the number of registrations.

Reply:
i see, but this invariant is more of an artifact of how we write our tests, not exactly how we expect the Queue to uphold in general.
I'd prefer to remove it as it's a little confusing.

Reply:
Ok, invariant removed in ed12203.